### DISCUSSION TITLE
National Freenet can prepare a safe uncensorable place for normal people

### DISCUSSION THREAD

**Discussion by PrinceOfSleeping** [WeAreMahsaAmini/Core#496] [Ideas]
As discussed here https://github.com/WeAreMahsaAmini/Core/discussions/6 a P2P solution is ultimately needed for Iran. Lack of communication among people is the best way to control them, which is why every social network is being censored, and government resorts to the so called national Internet (Intranet). While projects like https://github.com/tcfev/fordem are very promising, and I hope they flourish beyond our expectation, what I don't realize that is the need for making such projects from scratch. 

P2P is hard, let's not forget that; with limited resource, brains, energy and money, it is simply impossible.
1. Users are behind commercial NAT, UDP / TCP hole punching is needed to make P2P works. 
2. More than that, we have to deal with DHT bootstrapping problem. In other words, since users' IP are dynamic, a server with static IP is needed to share user's IP with other users.
3. Making communities (forums), messengers, file uploaders, etc in P2P is hard. Because unlike the centralized solution, there is actually no "place" to store these information, except for people's hard disks. Moreover, there has to be some preventive measures against data removal. For example, if a moderator of forum is arrested, the forum itself along with its information should survive. 
4. More importantly P2P by itself is not secure at all. If a user get caught in real world, based on the IPs that user was connected to, government can easily find other users. Being Anonymous is perhaps the hardest problem of P2P. I think this problem is present in fordem (and understandably so).

Now, a very mature project like https://freenetproject.org/ which is older than Iran's public Internet, addresses all these problems at the same time. You can read their papers if you want (and you probably should https://freenetproject.org/pages/about.html , even if you don't want to use it, it can give you lots of mature ideas), but the point is, why not using this open source project to build an ultimate app for people? If you are not familiar with freenet, you should probably read https://en.wikipedia.org/wiki/Freenet first.

Yes Freenet itself has to be modified for Iranian users. Why?! Well, because 
1. DHT bootstrapping problem. Freenet too, uses some initial servers to give a user other users' IP. Government has already blocked these initial servers. 
2. In completely National Internet, one cannot connect to Freenet because of its design. Freenet attempts to connect people all around the world. 
3. Freenet doesn't have an official client for smartphone devices, while people spend much of their time on smartphone devices anyway. 

So to address the problems above, I summarize my proposal:

1. It is an easy, just like how Tor uses bridges, we can have infinitely many initial servers that users can share with each other. Ideally, the code for these servers should be written in javascript so that it works in cheap shared host servers. More than that, we don't want to give away the identity of these servers' owners. Therefore, such servers should mimic the behavior of real websites and work at certain period of times. Besides, even if the server's owner is arrested, they can pretend that their server is hacked, and javascript code is injected to their websites, because of security issues. Ideally, when a government's website is hacked, the javascript code can be injected to their host for another good bridge as well! They can arrest themselves, in the name of betrayal or whatever.
User themselves, too, can act as servers, in a way that a user share other users IP with another users. Freenet does this, but not for initial connection.
2. Perhaps the hardest problem. The Freenet itself separates its network from the rest of internet. Yes, it's not a series of proxies like Tor. It is a separate network altogether (read their papers). We need a sub-network of Freenet's network for Iranian Users only. In the case that they resort to National Internet. Freenet has many active developers, and I am pretty sure they will be more than happy to either provide this feature themselves (they have already actually done something similar, that is called dark mode, but it's not a sub-network of freenet's network that works on national internet) or at least guide us where to start. https://github.com/freenet/fred/tree/build01494
3. Seems like other people have done this https://github.com/freenet-mobile/app .... though it's not an official client.

**Comment by arandomgstring** [comment]
@alazyprince 
I think you should've introduced freenet first, since it's very very rare to see someone who is already familiar with this protocol. Tor for sure, but freenet, just how many people do you think are aware of this?! But well whatever, they can check it in wikipedia if they desire so.

> In completely National Internet, one cannot connect to Freenet because of its design. Freenet attempts to connect people all around the world.

That's actually not quite right. Well, indeed Freenet does its best to connect you to "everywhere", but if it can't, you simply won't be able to open some Freenet pages. Unless your node has already done so. In other words, your second point doesn't stand. There is absolutely no need for any kind of modification. Freenet does work in national Internet! It's in its design! That said, it won't open any pages that national users don't have on their disks. When internet becomes intranet. And as you said, there is a client app for smartphone. So the only problem here, is your first point, which as you said is very easy to deal with.

@tcf-admin
I stand with @alazyprince point of view. Why do you want to reinvent the wheel from scratch when freenet have all features you are looking for? Writing a P2P app might not be impossible as per se, but it is indeed impractical. Moreover, besides 4 problems that @alazyprince mentioned, you have to maintain the security of your app against government. Do you think it's feasible, considering their manpower?! I think you should leave this heavy burden to Freenet. But you do you I guess.

If you use Freenet however, you would only need to maintain rendezvous servers and that is it. Making people familiar with this tool, making useful Iranian websites inside it, etc will become your goal, which is way easier. 

Another option that I'd like to add is I2P. You can use it too. Though unlike freenet, it doesn't support lots of features. There is a good comparison inside I2P itself https://geti2p.net/en/comparison/freenet .

**Comment by tcf-admin** [comment]
Hi @alazyprince & @arandomgstring, everyone else, 

Thank you very much for your contributions and for mentioning us.
At forDem we do several things but one thing that we do not do is the joyful act of reinventing the wheel.

Our P2P is not on IP, but on the radio level. What you two talked about (which is correct in its own sense) is not what we are trying to achieve. 
- We are combining a number of already existing & proven technologies to establish full connectivity among medium-sized groups using WiFi-Direct.
- The software of the platform's DHTs is more or less like any other functioning platform out there. But here again we're more interested in solving the hardware problems.
Part of our DHT carrying devices will be mobile, connected over a mesh over 3GPP, linked in some of the nodes via Ku band with internet providing satellites. (Fordem-physical is responsible for that)

If things go well these will be the highlights:
- We onion-route internet connections with the servers. 
- Servers stay linked together & things will be compatible w/ ActivityPub.
- We will have a stablecoin economy in the platform that incentivizes participation, on DHT, server and client parts.

**Reply by arandomgstring** [reply]
Sorry for misunderstanding. Your documentation page https://github.com/tcfev/forDem-documentation just returns 404 and explanations provided in here https://github.com/tcfev/fordem itself is honestly not very helpful, where can I read more about your model?

I presume by radio level, you meant ethernet level yes? Where devices communicate with each other using MAC addresses and private IPs that has nothing to do with ISPs and normal internet. 

Wouldn't your model becomes too "local" in a sense that a device cannot reach your "network" unless there exists another device nearby? (around 200m, at maximum in theory, I guess, but in practice, something like 45m or less). In other words, you will be having many parallel separated networks in each area that are not connected to each other, yes? And one of the routers, existing in each of these parallel networks has access to satellite internet?

To be honest, I see a lot of "features" in your explanation, but I still don't get it what you are trying to achieve. Let's forget about "features" that your network has to offer and focus on how devices connect to each other. 

You mentioned WiFi-Direct. Ok, a few close devices connect to each other using WiFi-Direct. How do they protect themselves from a malicious user who is trying to connect to them? After all, you cannot hide MAC address of yourself from devices that you are directly connected to. Do these users want to maintain their local connection, or they want to access a larger network (internet for example) at the same time?

**Reply by armantorkzaban** [reply]
Hello @arandomgstring, Arman here from [forDem](https://github.com/tcfev/fordem) team. Thanks for your reply and please excuse our delayed reply. 

You are right about fordem-documentation and we apologise for the inconvenience. The documentation is a work-in-progress & we are working on an update.

Below I will answer your questions and react to your assumptions (in an order other than how you had written them).

>To be honest, I see a lot of "features" in your explanation, but I still don't get it what you are trying to achieve. Let's forget about "features" that your network has to offer and focus on how devices connect to each other.

We do not have any extra features defined for the platform and we actively try to keep it as lean as possible. It only has the basic functionalities of a standard e-democracy platform, plus the ones users need to form a medium-sized WiFi-Direct mesh network. (which is expected from any platform made for places w/o IP access guarantee)

 > (around 200m, at maximum in theory, I guess, but in practice, something like 45m or less)
 
 Depending on the frequency used and the environmental factors the range can be different. For us at this phase the algorithm in charge of mesh network formation is important and variable WiFi range comes second.
 
> Do these users want to maintain their local connection, or they want to access a larger network (internet for example) at the same time?

Depends. We assume normal users without IP access, except for our mobile DHTs.

 >  In other words, you will be having many parallel separated networks in each area that are not connected to each other, yes?
 
Not necessarily. It depends on two factors. a) whether we will have our mobile DHTs in those networks or not & b) how mobile the users of a network will be & how successful their local synchronisation procedure has been, when they leave one network (and potentially enter another one)
If we have mobile DHTs available, they will try to bridge adjacent networks and even try to gain IP. If successful, we can expect something beyond a mid-size network.

>  How do they protect themselves from a malicious user who is trying to connect to them?

Discovery takes place with the help of the application. We use a set of unique identifiers, a certain protocol to exchange two things:
a. Values of certain size that we know the key of
b. limited number of key:values that we do not know beforehand (unique entries)

> you cannot hide MAC address of yourself from devices that you are directly connected to

No we can't. But users can change them if they want to go the extra mile. (And we teach them how to do so)
